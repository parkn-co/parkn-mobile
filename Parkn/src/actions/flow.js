export type Action = {
  type: String,
  payload: ?Any
};

export type AsyncAction = (dispatch: Function, getState: Function) => void
