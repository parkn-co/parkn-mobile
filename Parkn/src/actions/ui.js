// @flow
import type {Action} from 'flow-declarations/redux';

export const SWITCH_TO_VIEW = 'SWITCH_TO_VIEW';
export const switchToView = (payload: Object): Action => ({
  type: SWITCH_TO_VIEW,
  payload,
});
