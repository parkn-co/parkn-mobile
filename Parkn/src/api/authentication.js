// @flow
import type {UserForm} from 'flow-declarations/forms';
import Api from './base';

export function signIn({email, password}: UserForm): Promise<any> {
  return Api.post('auth/signin', {email, password});
}

export function signUp({email, firstName, lastName, password}: UserForm) {
  return Api.post('auth/signup', {email, firstName, lastName, password});
}

export function signOut(getState: Function): Promise<any> {
  return Api.authPost(getState, 'auth/signout');
}

export function fetchUser(getState: Function): Promise<any> {
  return Api.authGet(getState, 'users');
}
