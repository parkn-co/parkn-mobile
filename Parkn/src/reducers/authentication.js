// Import action handlers
import {
  AUTH_SET_FORM_VALUES,
  AUTH_SIGN_IN,
  AUTH_SET_ERRORS,
  AUTH_FETCH_USER,
  AUTH_IS_FETCHING,
} from '../actions/authentication';

import {REHYDRATE} from 'redux-persist/constants';

const initialState = {
  form: {
    firstName: '',
    lastName: '',
    email: '',
    password: '',
    confirmPassword: '',
  },
  formType: '',
  errors: {},
  token: null,
  user: {},
  isFetching: false,
};

// Action Handlers
const REDUCER_ACTION_HANDLERS = {
   [AUTH_SET_FORM_VALUES]: (state, {payload}) => ({
     ...state,
     form: {...state.form, ...payload},
     isFetching: false,
   }),
   [AUTH_SIGN_IN]: (state, {payload: {token}}) => ({
     ...state,
     token,
     isFetching: false,
   }),
   [AUTH_SET_ERRORS]: (state, {payload}) => ({
     ...state,
     errors: payload,
     isFetching: false,
   }),
   [AUTH_FETCH_USER]: (state, {payload}) => ({
     ...state,
     user: payload,
     isFetching: false,
   }),
   [AUTH_IS_FETCHING]: state => ({
     ...state,
     isFetching: true,
   }),
   [REHYDRATE]: (state, {payload}) => ({
     ...initialState,
     token: payload && payload.authentication && payload.authentication.token,
   }),
};

// Reducer
export default function authenticationReducer(state = initialState, action) {
  const handler = REDUCER_ACTION_HANDLERS[action.type]
  return handler ? handler(state, action) : state
}
