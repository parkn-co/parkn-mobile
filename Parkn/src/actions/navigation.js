// @flow
import {Action} from './flow';

export const NAV_NAVIGATE_TO = 'NAVIGATE_TO';
export function navigateTo(route: Object): Action {
  return {
    type: NAV_NAVIGATE_TO,
    payload: route,
  };
}

export const NAV_DID_NAVIGATE_TO = 'DID_NAVIGATE_TO';
export function didNavigateTo(route: Object): Action {
  return {
    type: NAV_DID_NAVIGATE_TO,
    payload: route,
  }
}

export const NAV_SET_IS_NAVIGATING = 'NAV_SET_IS_NAVIGATING';
export function setIsNavigating(isNavigating: Boolean): Action {
  return {
    type: NAV_SET_IS_NAVIGATING,
    payload: isNavigating,
  }
}
