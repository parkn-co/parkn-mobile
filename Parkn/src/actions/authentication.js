import * as AuthApi from '../api/authentication';

export const AUTH_IS_FETCHING = 'AUTH_IS_FETCHING';
export function setIsFetching() {
  return {
    type: AUTH_IS_FETCHING,
  };
}

export const AUTH_SET_FORM_VALUES = 'AUTH_SET_FORM_VALUES';
export function setFormValues(values) {
  return {
    type: AUTH_SET_FORM_VALUES,
    payload: values,
  };
}

export const AUTH_SIGN_IN = 'AUTH_SIGN_IN';
export const AUTH_SET_ERRORS = 'AUTH_SET_ERRORS';
export function authenticateWithValues(isSignUp, values) {
  return (dispatch, getState) => {
    dispatch({
      type: AUTH_SET_FORM_VALUES,
      payload: values
    });

    dispatch(setIsFetching());

    const {
      authentication: {form}
    } = getState();

    let method = 'signIn';
    if (isSignUp) {
      method = 'signUp';
    }

    AuthApi[method]({...form, ...values})
    .then(token => dispatch({
      type: AUTH_SIGN_IN,
      payload: token,
    }))
    .catch(err => {
      let payload = err;

      if (err.status === 409) {
        payload = {...err, email: err.error};
      } else if (err.status === 404 || err.status === 500) {
        payload = {...err, email: 'Incorrect email or password'};
      }

      dispatch({
        type: AUTH_SET_ERRORS,
        payload,
      })
    });
  }
}

export const AUTH_FETCH_USER = 'AUTH_FETCH_USER';
export function fetchUser() {
  return (dispatch, getState) => {
    dispatch(setIsFetching());

    AuthApi.fetchUser(getState)
    .then(token => dispatch({
      type: AUTH_FETCH_USER,
      payload: token,
    }))
    .catch(err => dispatch({
      type: AUTH_SET_ERRORS,
      payload: err,
    }));
  }
}
