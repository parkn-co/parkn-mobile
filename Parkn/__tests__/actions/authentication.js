/* eslint-env jest */

import 'react-native';
import React from 'react';
import { REHYDRATE } from 'redux-persist/constants';
import * as authenticationActions from '../../src/actions/authentication';

describe('Navigation actions', () => {
  it('return the correction action when setIsFetching', () => {
    expect(authenticationActions.setIsFetching()).toMatchSnapshot();
  });
  it('return the correction action when setFormValues', () => {
    expect(authenticationActions.setFormValues({ value: 'value' })).toMatchSnapshot();
  });
});
