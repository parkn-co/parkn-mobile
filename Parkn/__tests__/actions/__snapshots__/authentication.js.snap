exports[`Authentication actions return the correction action when setFormValues 1`] = `
Object {
  "payload": Object {
    "value": "value",
  },
  "type": "AUTH_SET_FORM_VALUES",
}
`;

exports[`Authentication actions return the correction action when setIsFetching 1`] = `
Object {
  "type": "AUTH_IS_FETCHING",
}
`;
